File: leetcode/41-first-missing-positive.rb
# @param {Integer[]} nums
# @return {Integer}
def first_missing_positive(nums)
  i, j = 0, 0
  while j < nums.size
    if nums[j] > 0
      nums[i] = nums[j]
      i += 1
    end
    j += 1
  end
  pos_count = i
  while i < j
    nums[i] = 0
    i += 1
  end
  i = 0
  while i < pos_count
    idx = nums[i].abs
    if idx <= nums.size and nums[idx-1] == 0
      nums[idx-1] = -1
    elsif idx <= nums.size and nums[idx-1] > 0
      nums[idx-1] = -nums[idx-1]
    end
    i += 1
  end
  i = 0
  while i < nums.size
    break if nums[i] >= 0
    i += 1
  end
  i + 1
end

File: leetcode/41-first-missing-positive.cpp
/*
Given an unsorted integer array, find the first missing positive integer.

For example,
Given [1,2,0] return 3,
and [3,4,-1,1] return 2.

Your algorithm should run in O(n) time and uses constant space.
*/

class Solution {
public:
    int firstMissingPositive(vector<int>& nums) {
        for (int i=0; i<nums.size(); ++i) {
            while (nums[i] > 0 && nums[i] != INT_MAX) {
                auto num = nums[i];
                if (num-1 > nums.size()) {
                    break;
                }
                if (nums[num-1] != INT_MAX) {
                    swap(nums[i], nums[num-1]);
                } else {
                    break;
                }
                nums[num-1] = INT_MAX;
            }
        }
        for (int i=0; i<nums.size(); ++i) {
            if (nums[i] != INT_MAX) {
                return i+1;
            }
        }
        return nums.size()+1;
    }
};
